libr/util/vector.c

```diff
@@ -16,155 +16,237 @@
-	}
-	int i;
-	for (i = vec->len; i > n; ) {
-		i--;
-		vec->a[last - first + i] = vec->a[i];
+
+
+R_API void *r_vector_insert(RVector *vec, size_t index, void *x) {
+	if (vec->len >= vec->capacity) {
+		RESIZE_OR_RETURN_NULL (NEXT_VECTOR_CAPACITY);
 	}
-	vec->len += last - first;
-	i = n;
-	while (first != last) {
-		vec->a[i++] = *first++;
+	void *p = r_vector_index_ptr (vec, index);
+	if (index < vec->len) {
+		memmove ((char *)p + vec->elem_size, p, vec->elem_size * (vec->len - index));
 	}
-	return &vec->a[n];
+	vec->len++;
+	r_vector_assign (vec, p, x);
+	return p;
 }
 
-R_API RVector *r_vector_new(void) {
-	return R_NEW0 (RVector);
+R_API void *r_vector_insert_range(RVector *vec, size_t index, void *first, size_t count) {
+	if (vec->len + count > vec->capacity) {
+		RESIZE_OR_RETURN_NULL (R_MAX (NEXT_VECTOR_CAPACITY, vec->len + count));
+	}
+	size_t sz = count * vec->elem_size;
+	void *p = r_vector_index_ptr (vec, index);
+	if (index < vec->len) {
+		memmove ((char *)p + sz, p, vec->elem_size * (vec->len - index));
+	}
+	vec->len += count;
+	memcpy (p, first, sz);
+	return p;
 }
 
-R_API void *r_vector_pop(RVector *vec) {
-	return vec->a[--vec->len];
+R_API void r_vector_pop(RVector *vec, void *into) {
+	if (into) {
+		r_vector_assign (vec, into, r_vector_index_ptr (vec, vec->len - 1));
+	}
+	vec->len--;
 }
 
-R_API void *r_vector_pop_front(RVector *vec) {
-	return r_vector_delete_at (vec, 0);
+R_API void r_vector_pop_front(RVector *vec, void *into) {
+	return r_vector_remove_at (vec, 0, into);
 }
 
-R_API void **r_vector_push(RVector *vec, void *x) {
+R_API void *r_vector_push(RVector *vec, void *x) {
 	if (vec->len >= vec->capacity) {
 		RESIZE_OR_RETURN_NULL (NEXT_VECTOR_CAPACITY);
 	}
-	vec->a[vec->len] = x;
-	return &vec->a[vec->len++];
+	return r_vector_assign_at (vec, vec->len++, x);
 }
 
-R_API void **r_vector_push_front(RVector *vec, void *x) {
+R_API void *r_vector_push_front(RVector *vec, void *x) {
 	return r_vector_insert (vec, 0, x);
 }
 
-R_API void **r_vector_reserve(RVector *vec, int capacity) {
+R_API void *r_vector_reserve(RVector *vec, size_t capacity) {
 	if (vec->capacity < capacity) {
 		RESIZE_OR_RETURN_NULL (capacity);
 	}
 	return vec->a;
 }
 
-R_API void **r_vector_shrink(RVector *vec) {
+R_API void *r_vector_shrink(RVector *vec) {
 	if (vec->len < vec->capacity) {
 		RESIZE_OR_RETURN_NULL (vec->len);
 	}
 	return vec->a;
 }
 
+
+
+
+
+static void pvector_free_elem(void *e, void *user) {
+	void *p = *((void **)e);
+	RPVectorFree elem_free = (RPVectorFree)user;
+	elem_free (p);
+}
+
+
+R_API void r_pvector_init(RPVector *vec, RPVectorFree free) {
+	r_vector_init (&vec->v, sizeof (void *));
+	vec->free = free;
+}
+
+R_API RPVector *r_pvector_new(RPVectorFree free) {
+	RPVector *v = R_NEW (RPVector);
+	if (!v) {
+		return NULL;
+	}
+	r_pvector_init (v, free);
+	return v;
+}
+
+R_API void r_pvector_clear(RPVector *vec) {
+	r_vector_clear (&vec->v, vec->free ? pvector_free_elem : NULL, vec->free);
```

Comment:
That's why the user pointer is needed. RVector can contain any kind of elements. It will call the free function on `vec->a + i * vec->elem_size`, i.e. a pointer to the element in the array. But for pvector, we know we store pointers, so it should be called on `*(vec->a + i * vec->elem_size)` instead. p_vector_free_elem just dereferences the given pointer and calles the function in user on it.